Add a BenchmarkDotNet project

JeremyKuhne opened the pull request:
This is to faciliate writing and running micro benchmarks.

Have to workaround a few issues to make this work:

1. Can't output to `artifacts/bin` as it contains a global.json (which breaks project finding logic and other things)
2. Need to be able to skip _CustomizeXlfSourceNames

Other things:

- Bump the BenchmarkDotNet version
- Add the Windows package versions to make it easy to use that functionality when needed

baronfel commented on build/GenerateResxSource.targets:
```diff
@@ -1,8 +1,13 @@
-          DependsOnTargets="_CustomizeResourceNames"
+          DependsOnTargets="$(CustomizeXlfSourceNamesDependsOn)"
           BeforeTargets="PrepareResourceNames;GatherXlf"
+          Condition="'$(SkipCustomizeXlfSourceNames)'!='true'"
```
You shouldn't need both this Condition and the change to the DependsOnTargets - if a Condition evaluates to `false` then the DependsOnTargets should not be run at all. Were you seeing different?

baronfel commented on eng/Versions.props:
```diff
@@ -140,7 +140,8 @@
     <MicrosoftDotNetInstallerWindowsSecurityTestDataPackageVersion>8.0.0-beta.23607.1</MicrosoftDotNetInstallerWindowsSecurityTestDataPackageVersion>
-    <BenchmarkDotNetPackageVersion>0.14.0</BenchmarkDotNetPackageVersion>
+    <BenchmarkDotNetPackageVersion>0.15.6</BenchmarkDotNetPackageVersion>
+    <BenchmarkDotNetDiagnosticsWindowsPackageVersion>0.14.0</BenchmarkDotNetDiagnosticsWindowsPackageVersion>
```
```suggestion
    <BenchmarkDotNetDiagnosticsWindowsPackageVersion>0.15.6</BenchmarkDotNetDiagnosticsWindowsPackageVersion>
```

If these packages version together we may want to use a single property to ensure they don't drift.

JeremyKuhne replied:
Not clear on why this was separate, I'll look into it and follow up when I get the chance.

baronfel reviewed (approved):
A few notes that may make maintenance/use simpler, but otherwise looks straightforward to me!

JeremyKuhne committed:
Add a BenchmarkDotNet project

This is to faciliate writing and running micro benchmarks.

Have to workaround a few issues to make this work:

1. Can't output to `artifacts/bin` as it contains a global.json (which breaks project finding logic and other things)
2. Need to be able to skip _CustomizeXlfSourceNames

Other things:

- Bump the BenchmarkDotNet version
- Add the Windows package versions to make it easy to use that functionality when needed

JeremyKuhne commented:
@MiYanni @baronfel updated per feedback.
